align_system/algorithms/alignment_adm_component.py

```diff
@@ -311,3 +311,147 @@ def _midpoint_eqn(self, kdma, opt_a_value, medical_delta, attr_delta):
+        """
+        if alignment_target is None:
+            raise RuntimeError(
+                "Assumption violated: `alignment_target` was None"
+            )
+
+        target_kdmas = alignment_target_to_attribute_targets(
+            alignment_target,
+            self.attributes)
+        target_kdmas = [dict(t) for t in target_kdmas]
+
+        choices = list(attribute_prediction_scores.keys())
+        if len(choices) != 2:
+            raise NotImplementedError("This alignment function has not yet been implemented for !=2 choices")
+
+        # Compute averages of predicted values
+        predictions = []
+        for choice, all_kdma_predictions in attribute_prediction_scores.items():
+            pred_dict = {"choice": choice}
+
+            # Get medical urgency
+            if med_urg_str not in all_kdma_predictions:
+                raise RuntimeError("Medical Urgency predictions required for this alignment function")
+
+            # Get KDMA predictions relevant to target
+            pred_dict.update(_get_avg_pred(all_kdma_predictions, target_kdmas))
+
+            predictions.append(pred_dict)
+
+        # Get relevance predictions relevant to target
+        if attribute_relevance is not None:
+            probe_relevance = _get_avg_pred(attribute_relevance, target_kdmas)
+        else:
+            probe_relevance = {target_kdma["kdma"]: 1.0 for target_kdma in target_kdmas}
+
+        # If more than 1 attribute is relevant, don't know what to do
+        relevant_kdmas = [kdma for kdma, relevance in probe_relevance.items() if relevance > 0]
+        if len(relevant_kdmas) != 1:
+            raise RuntimeError("This alignment function can only be used when 1 attribute is relevant")
+
+        # Guaranteed to only have 2 choices at this point due to earlier checks
+        opt_a, opt_b = predictions
+
+        for target_kdma in target_kdmas:
+            kdma = target_kdma["kdma"]
+            if kdma != relevant_kdmas[0]:
+                continue
+
+            intercept = None
+            medical_weight = None
+            attr_weight = None
+            if target_kdma["parameters"] is not None:
+                for param in target_kdma["parameters"]:
+                    if param["name"] == "intercept":
+                        intercept = param["value"]
+                    if param["name"] == "medical_weight":
+                        medical_weight = param["value"]
+                    if param["name"] == "attr_weight":
+                        attr_weight = param["value"]
+            if intercept is None or medical_weight is None or attr_weight is None:
+                raise RuntimeError("This alignment function requires an intercept, medical weight, and attr weight")
+
+            # PS medical for B should always be 0
+            # TODO: Should we explicitly only use opt_a for PS?
+            raw_medical_delta = opt_a[med_urg_str] - opt_b[med_urg_str]
```

Comment:
One of the more confusing pieces to me, is that the medical delta is used here, but for the attribute it's just the raw attribute score.  Looking over Amy's later messages (and her resolved example with z-scaling) and the slides it does seem like it's supposed to be the medical delta here, but confusing nonetheless.